### PR TITLE
perf(tree): cache canonical trie updates

### PR DESCRIPTION
**Pull request by rkrasiuk**
## Description

Cache trie updates for in-memory chains that extend the canonical one and reuse them on insert. 
In essence, this PR optimistically caches trie updates that were computed upon receiving new payload, provided it extends the canonical chain, and reuses these updates during subsequent forkchoice update.

Cache invalidation is very strict:
- if chain split occurs, trie updates are discarded
- if two chains are merged and either one doesn't have trie updates, the other updates are discarded.

_Pending automated and live testing_

### CODE REVIEW

**Review by mattsse — COMMENTED** [review]
this looks reasonable to me

pending @rakita

**Comment by gakonst** [comment]
Do we have any expected perf improvement here? What was measured / in which codepaths does this improve things?

**Comment by rkrasiuk** [comment]
@gakonst FCU processing is 90-100ms faster with caching. cache hit rate is pretty high, since most of the time the new payload will be committed next

File: crates/blockchain-tree/src/blockchain_tree.rs [hunk 1]
```diff
@@ -1160,29 +1160,43 @@ impl<DB: Database, EF: ExecutorFactory> BlockchainTree<DB, EF> {
...
-            return Err(RethError::Provider(ProviderError::StateRootMismatch(Box::new(
-                RootMismatch {
-                    root: GotExpected { got: state_root, expected: tip.state_root },
-                    block_number: tip.number,
-                    block_hash: tip.hash,
-                },
-            ))))
-        }
+        let (blocks, state, chain_trie_updates) = chain.into_inner();
+        let hashed_state = state.hash_state_slow();
```
Comment (rakita) [inline crates/blockchain-tree/src/blockchain_tree.rs:1164]: Should we call this only if there are no chain_trie_updates?

Comment (rakita) [reply crates/blockchain-tree/src/blockchain_tree.rs:1164]: Ah i see, theanswer is not. Maybe we can add those similar to how we added `chain_trie_updates` on the happy path, but that is for later.

**Review by rakita — APPROVED** [review]
lgtm

**Pull request merged**